### ISSUE TITLE
Multivariate Time Series Imputation with Generative Adversarial Networks

### ISSUE THREAD

**Issue by ghuijo** [yskim0/One-Day-One-Paper#100]
## Abstract
* propose to learn the overall distribution of a multivariate time series dataset with GAN, which is further used to generate the missing values for each sample
* A modified Gate Recurrent Unit is employed in GAN to model the temporal irregularity of the incomplete time series

## Introduction
* time series 데이터 문제점: the time intervals of the observations in time series are not always fixed
* PhysioNet dataset:  the maximum missing rate at each timestamp is always higher than 95%
* -> the processing of missing values in time series has become a very important problem
* 기존 imputation 방식 세 가지는 문제점 큼 & 기존 GAN은 complete training dataset을 필요로 함.
* In order to learn the latent relationships between observations with non-fixed time lags, we propose a novel RNN cell called GRUI which can take into account the non-fixed time lags and fade the influence of the past observations determined by the time lags.
* In the first phase, by adopting the GRUI in the discriminator and generator in GAN, the well trained adversarial model can learn the distribution of the whole dataset, the implicit relationships between observations and the temporal information of the dataset.
* In the second phase, we train the input “noise” of the generator of the GAN so that the generated time series is as close as possible to the original incomplete time series and the generated data’s probability of being real is the biggest.

## Methods
* time series X / introduce the mask matrix M
<image src="https://user-images.githubusercontent.com/67627471/130625554-ce5045a5-573f-4dd2-beb3-b1ac0ce93392.png" width="60%" height="60%" />

* In order to replace missing values in time series data with reasonable values, we first train a GAN based model to learn the distribution of the original time series dataset.
* fix the network structure and optimize the input random vector of the generator so that the generated fake time series can best replace the missing values
![image](https://user-images.githubusercontent.com/67627471/130626339-50ac2a85-7b37-4ce7-a09f-03b6ed3a3934.png)

* generator와 discriminator의 기본 모델로 GRU 사용.
* the time lags between two consecutive valid observations vary a lot due to data incompleteness
* -> traditional GRU cell이나 LSTM cell은 적합하지 않음.
*  The time lags between observations are very important since they follow an unknown nonuniform distribution
* propose the Gated Recurrent Unit for data Imputation (GRUI) cell to model the temporal irregularity of the incomplete time series
* calculation 방법과 계산된 δ 결과. (샘플)
![image](https://user-images.githubusercontent.com/67627471/130627257-8fdcee0c-fa2e-458d-ba22-31d19006999b.png)

* introduce a time decay vector β to control the influence of the past observations
![image](https://user-images.githubusercontent.com/67627471/130629073-b1263ea4-7c24-4441-82dc-b322a17b9354.png)

* use the negative exponential formulation to make sure that β_ti ∈ (0, 1]
*  in order to capture the interactions of the δ’s variables -> prefer full weight matrix to diagonal matrix for Wβ
* After we have got the decay vector -> update the GRU hidden state h_ti−1 by element-wise multiplying the decay factor β.
* Since we have used the batch normalization technology, the hidden state h is smaller than 1 with a high probability.
* We choose multiplicative decay way rather than some other decay ways such as hβ.
* GRUI의 update functions
![image](https://user-images.githubusercontent.com/67627471/130629990-968ef834-07bb-41d5-8b89-d83459c36cea.png)

* 기존 GRU와 본 논문의 GRUI 셀 비교
![image](https://user-images.githubusercontent.com/67627471/130630127-a03a4201-28df-4b77-9532-c6dff3ce8ff7.png)

* For any incomplete time series x, we try to find a best vector z from the latent input space so that the generated sample G(z) is most similar to x
  * introduce a way to measure the degree of imputation fitness.
  * define a two-part loss function to evaluate the fitness of imputation
  * The first part of this loss function is the masked reconstruction loss
  * The another part of this loss function is the discriminative loss
* Masked Reconstruction Loss
<image src="https://user-images.githubusercontent.com/67627471/130630699-213df373-28b1-414f-87f4-2734d18eff6d.png" width="50%" height="50%" />

* Discriminative Loss
<image src="https://user-images.githubusercontent.com/67627471/130630753-d2cfa215-57ff-4917-b4e8-1ce82f9280cf.png" width="30%" height="30%" />

* Imputation Loss
<image src="https://user-images.githubusercontent.com/67627471/130630791-a8c5c895-4f2f-4d4d-a654-5c1da6785895.png" width="50%" height="50%" />

*  replace the missing values of x with the generated G(z)
<image src="https://user-images.githubusercontent.com/67627471/130631076-2f2b408e-6a2c-404d-94e3-2b9f21380821.png" width="60%" height="60%" />

## datasets
* Physionet Challenge 2012 dataset (PhysioNet)
* KDD CUP 2018 Dataset (KDD)

## Results
<image src="https://user-images.githubusercontent.com/67627471/130631362-9717559a-4bd1-4cc9-a769-5fbce393e6c4.png" width="60%" height="60%" />

* mortality 예측 score
<image src="https://user-images.githubusercontent.com/67627471/130631459-0da70931-2ed4-4f00-8da4-dff8ad8751fb.png" width="60%" height="60%" />

* missing data imputation 방식에 따른 정확도 (MSE)
<image src="https://user-images.githubusercontent.com/67627471/130631596-4954ba50-5240-43a1-a5a9-c008768cf6bc.png" width="60%" height="60%" />

<image src="https://user-images.githubusercontent.com/67627471/130631682-b751601d-fc19-4625-af88-0fb2cff40d26.png" width="70%" height="70%" />

##
* 이거 구현한 코드 자세히 리뷰하기.

**Comment by ghuijo** [comment]
오오 제가 100번째 이슈 생성했네요!
